File: rust/src/wire/frame.rs
//! Brain wire-protocol frame codec.
//!
//! Re-implements Brain's 32-byte `BRN0` frame header. The conformance
//! corpus is the drift guard against the server.
//!
//! Layout — all multi-byte integers big-endian:
//! ```text
//! byte  0..4   magic "BRN0"
//! byte  4      version (= 1)
//! byte  5..7   opcode (u16 BE; high byte = namespace 0x00 / 0x01)
//! byte  7      flags (EOS=0x80, MPL=0x40, CMP=0x20)
//! byte  8..12  header_crc32c (u32 BE) — CRC32C over header bytes 0..8 ++ 12..32
//! byte 12..16  stream_id (u32 BE)
//! byte 16..19  payload_len (u24 BE)
//! byte 19      reserved_a (= 0)
//! byte 20..24  payload_crc32c (u32 BE; 0 when payload empty)
//! byte 24..32  reserved_b (8 bytes, = 0)
//! ```
//! Payload (not part of this header) = CBOR map + optional trailing raw
//! little-endian f32 vector section.

/// Magic bytes at offset 0: ASCII `"BRN0"`.
pub const MAGIC: [u8; 4] = *b"BRN0";

/// Wire-protocol version (header byte 4).
pub const WIRE_VERSION: u8 = 1;

/// Total header size, in bytes.
pub const HEADER_SIZE: usize = 32;

// Frame flag bits (header byte 7).
pub const FLAG_EOS: u8 = 0x80;
pub const FLAG_MPL: u8 = 0x40;
pub const FLAG_CMP: u8 = 0x20;

/// Flag bits that are defined; everything else MUST be zero on the wire.
/// A corrupt or hostile peer that sets a reserved bit is rejected rather
/// than silently reinterpreted.
pub const FLAGS_DEFINED_MASK: u8 = FLAG_EOS | FLAG_MPL | FLAG_CMP;

// Opcode namespaces (high byte of the u16 opcode).
pub const OPCODE_NS_SUBSTRATE: u8 = 0x00;
pub const OPCODE_NS_TYPED_GRAPH: u8 = 0x01;

// Field byte offsets within the 32-byte header.
pub const OFF_MAGIC: usize = 0;
pub const OFF_VERSION: usize = 4;
pub const OFF_OPCODE: usize = 5;
pub const OFF_FLAGS: usize = 7;
pub const OFF_HEADER_CRC: usize = 8;
pub const OFF_STREAM_ID: usize = 12;
pub const OFF_PAYLOAD_LEN: usize = 16;
pub const OFF_RESERVED_A: usize = 19;
pub const OFF_PAYLOAD_CRC: usize = 20;
pub const OFF_RESERVED_B: usize = 24;

/// Max payload size: `2^24 - 1` (the u24 `payload_len` field).
pub const MAX_PAYLOAD_BYTES: usize = 0xff_ff_ff;

/// Errors raised while decoding a frame off the wire. Each maps to a
/// wire protocol-error condition; a mismatch is fail-stop for the
/// connection (a corrupt frame desynchronizes the byte stream).
#[derive(Debug, thiserror::Error, PartialEq, Eq)]
pub enum FrameError {
    /// First four bytes were not `b"BRN0"`.
    #[error("bad frame magic")]
    BadMagic,
    /// Version byte did not match [`WIRE_VERSION`].
    #[error("unsupported wire version: got {got}, expected {expected}")]
    BadVersion { got: u8, expected: u8 },
    /// `header_crc32c` did not match the recomputed CRC.
    #[error("header CRC mismatch")]
    BadHeaderCrc,
    /// `payload_crc32c` did not match the CRC of the payload bytes.
    #[error("payload CRC mismatch")]
    BadPayloadCrc,
    /// `payload_len` exceeded the negotiated maximum — rejected before
    /// any payload buffer is reserved, so a hostile length can't force a
    /// large allocation.
    #[error("payload length {len} exceeds maximum {max}")]
    OversizePayload { len: u32, max: u32 },
    /// A reserved header byte or reserved flag bit was non-zero.
    #[error("reserved field or flag bit was non-zero")]
    ReservedNonZero,
    /// Not enough bytes buffered yet to decode the frame; caller should
    /// read more from the socket and retry.
    #[error("truncated frame: have {have} bytes, need {need}")]
    Truncated { have: usize, need: usize },
}

/// A decoded frame: header fields + the raw payload bytes (CBOR section
/// plus any trailing raw-vector section, undivided at this layer).
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Frame {
    pub opcode: u16,
    pub flags: u8,
    pub stream_id: u32,
    pub payload: Vec<u8>,
}

/// CRC32C over the header bytes that exclude the `header_crc32c` field
/// itself: bytes `0..8` concatenated with `12..32`. The 4-byte CRC slot
/// is held at zero so encode and decode agree on the covered bytes.
fn compute_header_crc(header: &[u8; HEADER_SIZE]) -> u32 {
    let mut buf = [0u8; HEADER_SIZE - 4];
    buf[..OFF_HEADER_CRC].copy_from_slice(&header[..OFF_HEADER_CRC]);
    buf[OFF_HEADER_CRC..].copy_from_slice(&header[OFF_STREAM_ID..]);
    crc32c::crc32c(&buf)
}

impl Frame {
    /// Build a frame from its parts.
    #[must_use]
    pub fn new(opcode: u16, flags: u8, stream_id: u32, payload: Vec<u8>) -> Self {
        Self {
            opcode,
            flags,
            stream_id,
            payload,
        }
    }

    /// Serialize the frame to a wire-ready byte vector: the 32-byte
    /// header with both CRC32C fields computed, followed by the payload.
    ///
    /// # Panics
    ///
    /// Panics if `payload.len()` exceeds [`MAX_PAYLOAD_BYTES`]; the u24
    /// length field physically cannot represent more.
    #[must_use]
    pub fn encode(&self) -> Vec<u8> {
        assert!(
            self.payload.len() <= MAX_PAYLOAD_BYTES,
            "payload length {} exceeds 24-bit max",
            self.payload.len()
        );

        let mut out = vec![0u8; HEADER_SIZE + self.payload.len()];
        {
            let header = &mut out[..HEADER_SIZE];
            header[OFF_MAGIC..OFF_MAGIC + 4].copy_from_slice(&MAGIC);
            header[OFF_VERSION] = WIRE_VERSION;
            header[OFF_OPCODE..OFF_OPCODE + 2].copy_from_slice(&self.opcode.to_be_bytes());
            header[OFF_FLAGS] = self.flags;
            // header_crc32c (8..12) left zero until computed below.
            header[OFF_STREAM_ID..OFF_STREAM_ID + 4].copy_from_slice(&self.stream_id.to_be_bytes());
            #[allow(clippy::cast_possible_truncation)]
            let len_be = (self.payload.len() as u32).to_be_bytes();
            header[OFF_PAYLOAD_LEN..OFF_PAYLOAD_LEN + 3].copy_from_slice(&len_be[1..]);
            // reserved_a (19) and reserved_b (24..32) left zero.
            // An empty payload carries a zero payload CRC by convention,
            // which crc32c of an empty slice already produces.
            let payload_crc = crc32c::crc32c(&self.payload);
            header[OFF_PAYLOAD_CRC..OFF_PAYLOAD_CRC + 4]
                .copy_from_slice(&payload_crc.to_be_bytes());
        }

        let header_arr: [u8; HEADER_SIZE] = out[..HEADER_SIZE]
            .try_into()
            .expect("invariant: out starts with exactly HEADER_SIZE header bytes");
        let header_crc = compute_header_crc(&header_arr);
        out[OFF_HEADER_CRC..OFF_HEADER_CRC + 4].copy_from_slice(&header_crc.to_be_bytes());

        out[HEADER_SIZE..].copy_from_slice(&self.payload);
        out
    }

    /// Decode a single frame off the front of `bytes`, returning the
    /// frame and the unconsumed tail so a caller can loop over a stream.
    ///
    /// Validates magic / version / reserved bytes / reserved flag bits,
    /// verifies the header CRC, then checks `payload_len` against
    /// [`MAX_PAYLOAD_BYTES`] **before** slicing the payload, and finally
    /// verifies the payload CRC.
    pub fn decode(bytes: &[u8]) -> Result<(Frame, &[u8]), FrameError> {
        if bytes.len() < HEADER_SIZE {
            return Err(FrameError::Truncated {
                have: bytes.len(),
                need: HEADER_SIZE,
            });
        }

        let header: [u8; HEADER_SIZE] = bytes[..HEADER_SIZE]
            .try_into()
            .expect("invariant: slice is exactly HEADER_SIZE bytes after the length check");

        if header[OFF_MAGIC..OFF_MAGIC + 4] != MAGIC {
            return Err(FrameError::BadMagic);
        }
        if header[OFF_VERSION] != WIRE_VERSION {
            return Err(FrameError::BadVersion {
                got: header[OFF_VERSION],
                expected: WIRE_VERSION,
            });
        }
        if header[OFF_RESERVED_A] != 0 || header[OFF_RESERVED_B..HEADER_SIZE] != [0u8; 8] {
            return Err(FrameError::ReservedNonZero);
        }
        let flags = header[OFF_FLAGS];
        if flags & !FLAGS_DEFINED_MASK != 0 {
            return Err(FrameError::ReservedNonZero);
        }

        let stored_header_crc = u32::from_be_bytes([
            header[OFF_HEADER_CRC],
            header[OFF_HEADER_CRC + 1],
            header[OFF_HEADER_CRC + 2],
            header[OFF_HEADER_CRC + 3],
        ]);
        if stored_header_crc != compute_header_crc(&header) {
            return Err(FrameError::BadHeaderCrc);
        }

        let payload_len = u32::from_be_bytes([
            0,
            header[OFF_PAYLOAD_LEN],
            header[OFF_PAYLOAD_LEN + 1],
            header[OFF_PAYLOAD_LEN + 2],
        ]);
        // Reject an over-long claim before reserving any payload buffer —
        // a corrupt or hostile length must not drive a large allocation.
        #[allow(clippy::cast_possible_truncation)]
        if payload_len as usize > MAX_PAYLOAD_BYTES {
            return Err(FrameError::OversizePayload {
                len: payload_len,
                max: MAX_PAYLOAD_BYTES as u32,
            });
        }

        let need = HEADER_SIZE + payload_len as usize;
        if bytes.len() < need {
            return Err(FrameError::Truncated {
                have: bytes.len(),
                need,
            });
        }
        let payload_slice = &bytes[HEADER_SIZE..need];

        let stored_payload_crc = u32::from_be_bytes([
            header[OFF_PAYLOAD_CRC],
            header[OFF_PAYLOAD_CRC + 1],
            header[OFF_PAYLOAD_CRC + 2],
            header[OFF_PAYLOAD_CRC + 3],
        ]);
        if stored_payload_crc != crc32c::crc32c(payload_slice) {
            return Err(FrameError::BadPayloadCrc);
        }

        let frame = Frame {
            opcode: u16::from_be_bytes([header[OFF_OPCODE], header[OFF_OPCODE + 1]]),
            flags,
            stream_id: u32::from_be_bytes([
                header[OFF_STREAM_ID],
                header[OFF_STREAM_ID + 1],
                header[OFF_STREAM_ID + 2],
                header[OFF_STREAM_ID + 3],
            ]),
            payload: payload_slice.to_vec(),
        };
        Ok((frame, &bytes[need..]))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample() -> Frame {
        Frame::new(0x0021, FLAG_EOS, 7, b"hello brain".to_vec())
    }

    #[test]
    fn encode_decode_round_trip() {
        let original = sample();
        let bytes = original.encode();
        assert_eq!(bytes.len(), HEADER_SIZE + original.payload.len());
        let (decoded, rest) = Frame::decode(&bytes).expect("decode");
        assert!(rest.is_empty());
        assert_eq!(decoded, original);
    }

    #[test]
    fn empty_payload_has_zero_payload_crc() {
        let frame = Frame::new(0x0010, 0, 0, Vec::new());
        let bytes = frame.encode();
        assert_eq!(bytes.len(), HEADER_SIZE);
        assert_eq!(&bytes[OFF_PAYLOAD_CRC..OFF_PAYLOAD_CRC + 4], &[0u8; 4]);
        let (decoded, rest) = Frame::decode(&bytes).expect("decode empty");
        assert!(rest.is_empty());
        assert!(decoded.payload.is_empty());
    }

    #[test]
    fn decode_returns_unconsumed_tail() {
        let mut bytes = sample().encode();
        bytes.extend_from_slice(b"TAIL");
        let (_frame, rest) = Frame::decode(&bytes).expect("decode");
        assert_eq!(rest, b"TAIL");
    }

    #[test]
    fn rejects_bad_magic() {
        let mut bytes = sample().encode();
        bytes[0] = b'X';
        assert_eq!(Frame::decode(&bytes), Err(FrameError::BadMagic));
    }

    #[test]
    fn rejects_bad_version() {
        let mut bytes = sample().encode();
        bytes[OFF_VERSION] = 99;
        assert_eq!(
            Frame::decode(&bytes),
            Err(FrameError::BadVersion {
                got: 99,
                expected: 1
            })
        );
    }

    #[test]
    fn rejects_bad_header_crc() {
        let mut bytes = sample().encode();
        bytes[OFF_HEADER_CRC] ^= 0xFF;
        assert_eq!(Frame::decode(&bytes), Err(FrameError::BadHeaderCrc));
    }

    #[test]
    fn rejects_bad_payload_crc() {
        let mut bytes = sample().encode();
        bytes[HEADER_SIZE] ^= 0xFF;
        assert_eq!(Frame::decode(&bytes), Err(FrameError::BadPayloadCrc));
    }

    #[test]
    fn rejects_reserved_a_nonzero() {
        let mut bytes = sample().encode();
        bytes[OFF_RESERVED_A] = 1;
        // Re-seal the header CRC so the reserved check (not the CRC
        // check) is the one that fires.
        let mut header: [u8; HEADER_SIZE] = bytes[..HEADER_SIZE].try_into().unwrap();
        let crc = compute_header_crc(&{
            let mut h = header;
            h[OFF_HEADER_CRC..OFF_HEADER_CRC + 4].copy_from_slice(&[0; 4]);
            h
        });
        header[OFF_HEADER_CRC..OFF_HEADER_CRC + 4].copy_from_slice(&crc.to_be_bytes());
        bytes[..HEADER_SIZE].copy_from_slice(&header);
        assert_eq!(Frame::decode(&bytes), Err(FrameError::ReservedNonZero));
    }

    #[test]
    fn rejects_reserved_flag_bit() {
        // 0x01 is a reserved flag bit. Build the frame raw so the header
        // CRC covers the bad flag, isolating the reserved-flag check.
        let frame = Frame::new(0x0010, 0x01, 0, Vec::new());
        let bytes = frame.encode();
        assert_eq!(Frame::decode(&bytes), Err(FrameError::ReservedNonZero));
    }

    #[test]
    fn rejects_truncated_header() {
        let bytes = sample().encode();
        let truncated = &bytes[..10];
        assert_eq!(
            Frame::decode(truncated),
            Err(FrameError::Truncated {
                have: 10,
                need: HEADER_SIZE
            })
        );
    }

    #[test]
    fn rejects_truncated_payload() {
        let bytes = sample().encode();
        let truncated = &bytes[..bytes.len() - 3];
        assert!(matches!(
            Frame::decode(truncated),
            Err(FrameError::Truncated { .. })
        ));
    }

    /// Standard CRC32C check vector pins the polynomial choice.
    #[test]
    fn crc32c_check_vector() {
        assert_eq!(crc32c::crc32c(b"123456789"), 0xE306_9283);
    }
}

File: typescript/src/wire/frame.ts
/**
 * Wire protocol: BRN0 framing and CBOR payload codec.
 *
 * All multi-byte header fields are **big-endian**. Payloads are CBOR with a
 * fixed per-opcode field order (reproducible-deterministic, not bytewise
 * key-sort).
 *
 * Cross-language gotchas this codec MUST honor (how Brain's corpus encodes
 * values, and where naive TS ports go wrong):
 *
 * - IDs / `[u8; N]` / UUIDs are CBOR **byte strings** (major type 2) ->
 *   `Uint8Array`. NOT number[] arrays, NOT hex strings, NOT hyphenated-UUID
 *   text. (Pick a CBOR lib that distinguishes byte strings from arrays — e.g.
 *   `cbor-x` with `useRecords:false` / tagged Uint8Array, or `cborg`.)
 * - Enums (memory `kind`, `ForgetMode`, `AuthMethod`) are **integers**.
 * - f32 fields round-trip through 32-bit floats (`Math.fround`), not JS's
 *   native 64-bit numbers.
 *
 * The conformance corpus is the drift guard: every `frame_*` golden must
 * re-encode byte-for-byte and decode back to its header fields.
 */

import { crc32c } from "./crc32c.js";

export { crc32c };

// ---- Header layout (32 bytes, fixed, big-endian) ---------------------------

export const HEADER_LEN = 32 as const;
/** Frame magic / start sentinel: ASCII "BRN0". */
export const MAGIC: Uint8Array = new Uint8Array([0x42, 0x52, 0x4e, 0x30]);
export const VERSION = 1 as const;
export const MAX_PAYLOAD_LEN = (1 << 24) - 1;

// Field offsets into the 32-byte header.
export const OFF_MAGIC = 0; // bytes 0..4
export const OFF_VERSION = 4; // byte 4
export const OFF_OPCODE = 5; // bytes 5..7  (u16 BE)
export const OFF_FLAGS = 7; // byte 7
export const OFF_HEADER_CRC = 8; // bytes 8..12 (u32 BE)
export const OFF_STREAM_ID = 12; // bytes 12..16 (u32 BE)
export const OFF_PAYLOAD_LEN = 16; // bytes 16..19 (u24 BE)
export const OFF_RESERVED_A = 19; // byte 19 (MUST be zero)
export const OFF_PAYLOAD_CRC = 20; // bytes 20..24 (u32 BE)
export const OFF_RESERVED_B = 24; // bytes 24..32 (MUST be zero)

// ---- Frame flags -----------------------------------------------------------

export const FLAG_EOS = 0x80; // end-of-stream: last frame for this streamId
export const FLAG_MPL = 0x40; // multi-payload: more frames follow in this message
export const FLAG_CMP = 0x20; // compressed payload (zstd)
export const FLAGS_RESERVED_MASK = 0x1f; // low five bits reserved; MUST be zero

// The header CRC is computed over header bytes 0..8 ++ 12..32 (its own four
// bytes at 8..12 are excluded). The payload CRC is over the payload bytes and
// MUST be zero when the payload is empty. Both are CRC32C (Castagnoli).

/** Flag bits that are defined; any other bit set on the wire is rejected. */
export const FLAGS_DEFINED_MASK = FLAG_EOS | FLAG_MPL | FLAG_CMP;

/** A decoded BRN0 frame. */
export interface Frame {
  opcode: number;
  streamId: number;
  flags: number;
  payload: Uint8Array;
}

/**
 * Failures decoding a frame off the wire. A mismatch is fail-stop for the
 * connection: a corrupt frame desynchronizes the byte stream, so the caller
 * cannot safely resume from the next byte.
 */
export type FrameErrorKind =
  | "BadMagic"
  | "BadVersion"
  | "BadHeaderCrc"
  | "BadPayloadCrc"
  | "OversizePayload"
  | "ReservedNonZero"
  | "Truncated";

export class FrameError extends Error {
  constructor(
    public readonly kind: FrameErrorKind,
    message: string,
  ) {
    super(message);
    this.name = "FrameError";
  }
}

// The header CRC covers the header bytes that exclude the CRC field itself:
// bytes 0..8 concatenated with 12..32. Encode and decode agree on the covered
// range by holding the 4-byte CRC slot at zero while computing it.
function computeHeaderCrc(header: Uint8Array): number {
  const buf = new Uint8Array(HEADER_LEN - 4);
  buf.set(header.subarray(0, OFF_HEADER_CRC), 0);
  buf.set(header.subarray(OFF_STREAM_ID, HEADER_LEN), OFF_HEADER_CRC);
  return crc32c(buf);
}

function writeU32Be(buf: Uint8Array, offset: number, value: number): void {
  buf[offset] = (value >>> 24) & 0xff;
  buf[offset + 1] = (value >>> 16) & 0xff;
  buf[offset + 2] = (value >>> 8) & 0xff;
  buf[offset + 3] = value & 0xff;
}

function readU32Be(buf: Uint8Array, offset: number): number {
  return (
    ((buf[offset]! << 24) |
      (buf[offset + 1]! << 16) |
      (buf[offset + 2]! << 8) |
      buf[offset + 3]!) >>>
    0
  );
}

/**
 * Encode a complete BRN0 frame (header ++ payload).
 *
 * Recomputes payloadLen, payloadCrc32c and headerCrc32c, then returns the
 * 32-byte header followed by the payload bytes. An empty payload carries a zero
 * payload CRC, which `crc32c` of an empty slice already produces.
 */
export function encodeFrame(
  opcode: number,
  streamId: number,
  flags: number,
  payload: Uint8Array,
): Uint8Array {
  if (payload.length > MAX_PAYLOAD_LEN) {
    throw new FrameError(
      "OversizePayload",
      `payload length ${payload.length} exceeds 24-bit max ${MAX_PAYLOAD_LEN}`,
    );
  }

  const out = new Uint8Array(HEADER_LEN + payload.length);
  out.set(MAGIC, OFF_MAGIC);
  out[OFF_VERSION] = VERSION;
  out[OFF_OPCODE] = (opcode >>> 8) & 0xff;
  out[OFF_OPCODE + 1] = opcode & 0xff;
  out[OFF_FLAGS] = flags & 0xff;
  // header_crc32c (8..12) stays zero until computed below.
  writeU32Be(out, OFF_STREAM_ID, streamId >>> 0);
  // payload_len is a u24: write the low three bytes of a big-endian u32.
  out[OFF_PAYLOAD_LEN] = (payload.length >>> 16) & 0xff;
  out[OFF_PAYLOAD_LEN + 1] = (payload.length >>> 8) & 0xff;
  out[OFF_PAYLOAD_LEN + 2] = payload.length & 0xff;
  // reserved_a (19) and reserved_b (24..32) stay zero.
  writeU32Be(out, OFF_PAYLOAD_CRC, crc32c(payload));

  const headerCrc = computeHeaderCrc(out.subarray(0, HEADER_LEN));
  writeU32Be(out, OFF_HEADER_CRC, headerCrc);

  out.set(payload, HEADER_LEN);
  return out;
}

/**
 * Decode one BRN0 frame from the front of `buf`, returning the frame and the
 * unconsumed tail (so callers can loop over a stream of frames).
 *
 * Verifies magic, version, both CRC32Cs and the reserved bits. The max-payload
 * check happens before the payload is read, so a hostile length field cannot
 * drive a large allocation.
 */
export function decodeFrame(buf: Uint8Array): { frame: Frame; rest: Uint8Array } {
  if (buf.length < HEADER_LEN) {
    throw new FrameError(
      "Truncated",
      `truncated frame: have ${buf.length} bytes, need ${HEADER_LEN}`,
    );
  }

  const header = buf.subarray(0, HEADER_LEN);

  for (let i = 0; i < MAGIC.length; i++) {
    if (header[OFF_MAGIC + i] !== MAGIC[i]) {
      throw new FrameError("BadMagic", "bad frame magic");
    }
  }
  if (header[OFF_VERSION] !== VERSION) {
    throw new FrameError(
      "BadVersion",
      `unsupported wire version: got ${header[OFF_VERSION]}, expected ${VERSION}`,
    );
  }

  // reserved_a and the eight reserved_b bytes must all be zero.
  if (header[OFF_RESERVED_A] !== 0) {
    throw new FrameError("ReservedNonZero", "reserved field was non-zero");
  }
  for (let i = OFF_RESERVED_B; i < HEADER_LEN; i++) {
    if (header[i] !== 0) {
      throw new FrameError("ReservedNonZero", "reserved field was non-zero");
    }
  }

  const flags = header[OFF_FLAGS]!;
  if ((flags & ~FLAGS_DEFINED_MASK) !== 0) {
    throw new FrameError("ReservedNonZero", "reserved flag bit was non-zero");
  }

  const storedHeaderCrc = readU32Be(header, OFF_HEADER_CRC);
  if (storedHeaderCrc !== computeHeaderCrc(header)) {
    throw new FrameError("BadHeaderCrc", "header CRC mismatch");
  }

  const payloadLen =
    (header[OFF_PAYLOAD_LEN]! << 16) |
    (header[OFF_PAYLOAD_LEN + 1]! << 8) |
    header[OFF_PAYLOAD_LEN + 2]!;
  // Reject an over-long claim before slicing any payload buffer.
  if (payloadLen > MAX_PAYLOAD_LEN) {
    throw new FrameError(
      "OversizePayload",
      `payload length ${payloadLen} exceeds maximum ${MAX_PAYLOAD_LEN}`,
    );
  }

  const need = HEADER_LEN + payloadLen;
  if (buf.length < need) {
    throw new FrameError("Truncated", `truncated frame: have ${buf.length} bytes, need ${need}`);
  }

  const payload = buf.slice(HEADER_LEN, need);
  const storedPayloadCrc = readU32Be(header, OFF_PAYLOAD_CRC);
  if (storedPayloadCrc !== crc32c(payload)) {
    throw new FrameError("BadPayloadCrc", "payload CRC mismatch");
  }

  const frame: Frame = {
    opcode: ((header[OFF_OPCODE]! << 8) | header[OFF_OPCODE + 1]!) >>> 0,
    streamId: readU32Be(header, OFF_STREAM_ID),
    flags,
    payload,
  };
  return { frame, rest: buf.slice(need) };
}
